dolphinscheduler-storage-plugin/dolphinscheduler-storage-gcs/src/main/java/org/apache/dolphinscheduler/plugin/storage/gcs/GcsStorageOperator.java

```diff
@@ -0,0 +1,494 @@
+
+    }
+
+    public void init() {
+        try {
+            credential = readCredentials();
+            bucketName = readBucketName();
+            gcsStorage = buildGcsStorage(credential);
+
+            checkBucketNameExists(bucketName);
+        } catch (IOException e) {
+            log.error("GCS Storage operator init failed", e);
+        }
+    }
+
+    protected Storage buildGcsStorage(String credential) throws IOException {
+        return StorageOptions.newBuilder()
+                .setCredentials(ServiceAccountCredentials.fromStream(
+                        Files.newInputStream(Paths.get(credential))))
+                .build()
+                .getService();
+    }
+
+    protected String readCredentials() {
+        return PropertyUtils.getString(Constants.GOOGLE_CLOUD_STORAGE_CREDENTIAL);
+    }
+
+    protected String readBucketName() {
+        return PropertyUtils.getString(Constants.GOOGLE_CLOUD_STORAGE_BUCKET_NAME);
+    }
+
+    @Override
+    public void createTenantDirIfNotExists(String tenantCode) throws Exception {
+        mkdir(tenantCode, getGcsResDir(tenantCode));
+        mkdir(tenantCode, getGcsUdfDir(tenantCode));
+    }
+
+    @Override
+    public String getResDir(String tenantCode) {
+        return getGcsResDir(tenantCode) + FOLDER_SEPARATOR;
+    }
+
+    @Override
+    public String getUdfDir(String tenantCode) {
+        return getGcsUdfDir(tenantCode) + FOLDER_SEPARATOR;
+    }
+
+    @Override
+    public String getResourceFileName(String tenantCode, String fileName) {
+        if (fileName.startsWith(FOLDER_SEPARATOR)) {
+            fileName.replaceFirst(FOLDER_SEPARATOR, EMPTY_STRING);
+        }
+        return String.format(FORMAT_S_S, getGcsResDir(tenantCode), fileName);
+    }
+
+    @Override
+    public String getResourceFileName(String fullName) {
+        return null;
+    }
+
+    @Override
+    public String getFileName(ResourceType resourceType, String tenantCode, String fileName) {
+        if (fileName.startsWith(FOLDER_SEPARATOR)) {
+            fileName = fileName.replaceFirst(FOLDER_SEPARATOR, EMPTY_STRING);
+        }
+        return getDir(resourceType, tenantCode) + fileName;
+    }
+
+    @Override
+    public void download(String tenantCode, String srcFilePath, String dstFilePath, boolean deleteSource,
+                         boolean overwrite) throws IOException {
+        File dstFile = new File(dstFilePath);
+        if (dstFile.isDirectory()) {
+            Files.delete(dstFile.toPath());
+        } else {
+            Files.createDirectories(dstFile.getParentFile().toPath());
+        }
+
+        Blob blob = gcsStorage.get(BlobId.of(bucketName, srcFilePath));
+        blob.downloadTo(Paths.get(dstFilePath));
+    }
+
+    @Override
+    public boolean exists(String fullName) throws IOException {
+        return isObjectExists(fullName);
+    }
+
+    @Override
+    public boolean delete(String filePath, boolean recursive) throws IOException {
```

Comment:
it seems the parameter `recursive` not use in this method, so GCS does not support delete recursive in this method?

Reply:
Hi, @zhongjiajie , thanks a lot for your kind review. Perhaps for some historical reasons, `recursive` is not used.

Another interface below is called when deleting recursively (also in the `S3StorageOperator` and `OssStorageOperator`). 

```
@Override
    public boolean delete(String fullName, List<String> childrenPathList, boolean recursive) throws IOException {
        // append the resource fullName to the list for deletion.
        childrenPathList.add(fullName);

        boolean result = true;
        for (String filePath : childrenPathList) {
            if (!delete(filePath, recursive)) {
                result = false;
            }
        }

        return result;
    }
```